Dip/Dip/Definition.swift

```diff
@@ -50,20 +50,62 @@ func ==(lhs: DefinitionKey, rhs: DefinitionKey) -> Bool {
 public enum ComponentScope {
   /// Indicates that a new instance of the component will be created each time it's resolved.
   case Prototype
+  /// Indicates that instances will be reused during resolve but will be discurded when topmost `resolve` method returns.
+  case ObjectGraph
   /// Indicates that resolved component should be retained by container and always reused.
   case Singleton
 }
 
 ///Definition of type T describes how instances of this type should be created when they are resolved by container.
 public final class DefinitionOf<T>: Definition {
```

Comment:
Yeah I agree and I do care about being able to use the trailing closure syntax, feels nice there.

I think I'm ok with:
* Removing `inScope` method and instead make it a parameter on `register`
* Either make `DefinitionOf` a `struct` + make `resolveDependencies` `mutating`
* Or keep `DefinitionOf` a `final class`
* In either case, make it `fatalError` if `resolveDependenciesBlock` is already non-`nil` when you call it, meaning that it's the second time you call it
* In such context, I'm wondering if it's still worth it to `return self` there as you won't probably need chaining with that new configuration, so better not tempt the user to chain somehow (besides, I don't like chaining mutating functions, I think mutating functions should be their own instructions)

What do you think?